File: engine/platform/macos-reference/MotionBlurQualityPlanner.swift
import Foundation

struct MotionBlurSamplePlan {
    let samples: [(time: Double, weight: Double)]
    let maxPixelDisplacement: Double
    let sampleCount: Int
}

enum MotionBlurQualityPlanner {
    static func plan(
        pass: FXPassNode,
        node: RenderGraphNode,
        frameTime: Double,
        fps: Double,
        maxSamples: Int,
        nodeAtTime: (Double) -> RenderGraphNode?
    ) -> MotionBlurSamplePlan {
        let params = pass.params.objectValue ?? [:]
        let safeFPS = max(1, fps)
        let frameDuration = 1 / safeFPS
        let shutterAngle = min(1440, max(0, params["shutterAngle"]?.numberValue ?? (params["shutter"]?.numberValue ?? 1) * 180))
        let shutterPhase = min(720, max(-720, params["shutterPhase"]?.numberValue ?? -shutterAngle / 2))
        let amount = min(10, max(0, params["amount"]?.numberValue ?? params["strength"]?.numberValue ?? 1))
        let shutterDuration = frameDuration * shutterAngle / 360 * amount
        let centerTime = frameTime + frameDuration * shutterPhase / 360
        let start = max(0, centerTime - shutterDuration / 2)
        let end = max(start, centerTime + shutterDuration / 2)

        let requested = requestedSampleCount(params: params)
        let displacement = maxPixelDisplacement(
            reference: node,
            times: [start, start + shutterDuration * 0.25, centerTime, start + shutterDuration * 0.75, end],
            nodeAtTime: nodeAtTime
        )
        let adaptive = adaptiveSampleCount(displacement: displacement, shutterAngle: shutterAngle, amount: amount)
        let sampleCount = min(maxSamples, max(requested, adaptive, 2))
        let curve = params["sampleCurve"]?.stringValue ?? "filmic"
        let rawSamples = (0..<sampleCount).map { index -> (time: Double, weight: Double) in
            let progress = sampleCount == 1 ? 0.5 : Double(index) / Double(sampleCount - 1)
            let time = start + progress * (end - start)
            return (time, reconstructionWeight(progress: progress, curve: curve))
        }
        let total = max(0.0001, rawSamples.reduce(0) { $0 + $1.weight })
        return MotionBlurSamplePlan(
            samples: rawSamples.map { ($0.time, $0.weight / total) },
            maxPixelDisplacement: displacement,
            sampleCount: sampleCount
        )
    }

    private static func requestedSampleCount(params: [String: JSONValue]) -> Int {
        if let value = params["quality"]?.stringValue {
            switch value {
            case "draft": return 8
            case "preview": return 16
            case "high": return 48
            case "cinematic", "best", "ultra": return 96
            default: break
            }
        }
        return max(16, Int((params["samples"]?.numberValue ?? 24).rounded()))
    }

    private static func adaptiveSampleCount(displacement: Double, shutterAngle: Double, amount: Double) -> Int {
        let pixelsPerSample = 1.35
        let displacementSamples = Int(ceil(max(0, displacement) / pixelsPerSample))
        let shutterSamples = Int(ceil(max(0, shutterAngle * max(1, amount)) / 18))
        return max(displacementSamples, shutterSamples)
    }

    private static func maxPixelDisplacement(
        reference: RenderGraphNode,
        times: [Double],
        nodeAtTime: (Double) -> RenderGraphNode?
    ) -> Double {
        let referenceCorners = layerCorners(reference.transform)
        var displacement = 0.0
        for time in times {
            guard let sample = nodeAtTime(max(0, time)) else { continue }
            let sampleCorners = layerCorners(sample.transform)
            for index in 0..<min(referenceCorners.count, sampleCorners.count) {
                displacement = max(displacement, hypot(sampleCorners[index].x - referenceCorners[index].x, sampleCorners[index].y - referenceCorners[index].y))
            }
        }
        return displacement
    }

    private static func layerCorners(_ transform: RenderTransform) -> [(x: Double, y: Double)] {
        let centerX = transform.x + transform.width * transform.anchorX
        let centerY = transform.y + transform.height * transform.anchorY
        let left = -transform.width * transform.anchorX
        let right = transform.width * (1 - transform.anchorX)
        let top = -transform.height * transform.anchorY
        let bottom = transform.height * (1 - transform.anchorY)
        let radians = transform.rotationDegrees * .pi / 180
        let cosValue = cos(radians)
        let sinValue = sin(radians)
        return [
            point(localX: left, localY: top, transform: transform, centerX: centerX, centerY: centerY, cosValue: cosValue, sinValue: sinValue),
            point(localX: right, localY: top, transform: transform, centerX: centerX, centerY: centerY, cosValue: cosValue, sinValue: sinValue),
            point(localX: right, localY: bottom, transform: transform, centerX: centerX, centerY: centerY, cosValue: cosValue, sinValue: sinValue),
            point(localX: left, localY: bottom, transform: transform, centerX: centerX, centerY: centerY, cosValue: cosValue, sinValue: sinValue)
        ]
    }

    private static func point(localX: Double, localY: Double, transform: RenderTransform, centerX: Double, centerY: Double, cosValue: Double, sinValue: Double) -> (x: Double, y: Double) {
        let skewedX = localX + tan(transform.skewXDegrees * .pi / 180) * localY
        let skewedY = tan(transform.skewYDegrees * .pi / 180) * localX + localY
        let scaledX = skewedX * transform.scaleX
        let scaledY = skewedY * transform.scaleY
        return (
            x: centerX + scaledX * cosValue - scaledY * sinValue,
            y: centerY + scaledX * sinValue + scaledY * cosValue
        )
    }

    private static func reconstructionWeight(progress: Double, curve: String) -> Double {
        switch curve {
        case "uniform":
            return 1
        case "centerWeighted":
            let distance = abs(progress - 0.5) * 2
            return max(0.01, 1 - distance * 0.55)
        default:
            let a0 = 0.35875
            let a1 = 0.48829
            let a2 = 0.14128
            let a3 = 0.01168
            let x = 2 * .pi * progress
            return max(0.001, a0 - a1 * cos(x) + a2 * cos(2 * x) - a3 * cos(3 * x))
        }
    }
}

File: src/core/hyperframe/fx/motionBlurQualityPlanner.ts
import type { HyperFrameEffectInstance, HyperFrameFrameDescriptorLayer } from "../contracts/types";
import type { FrameQualityPolicy } from "../render-plan/qualityPolicy";

export type MotionBlurQualityTier = "off" | "draft" | "preview" | "high" | "cinematic" | "clamped";

export type MotionBlurSamplePlan = {
  version: 2;
  shutterStartTime: number;
  shutterEndTime: number;
  sampleTimes: number[];
  weights: number[];
  estimatedMaxPixelDisplacement: number;
  estimatedAngularSweep: number;
  estimatedEdgeVelocity: number;
  requiredSampleCount: number;
  selectedSampleCount: number;
  qualityTier: MotionBlurQualityTier;
  diagnostics: string[];
};

const clamp = (value: number, min: number, max: number) => (
  Math.max(min, Math.min(max, value))
);

const toFiniteNumber = (value: unknown, fallback: number) => {
  const next = Number(value);
  return Number.isFinite(next) ? next : fallback;
};

const requestedSamples = (effect: HyperFrameEffectInstance, fallback: number) => {
  const quality = String(effect.params.quality || effect.quality || "auto");
  if (quality === "draft") return 8;
  if (quality === "preview") return 16;
  if (quality === "high") return 48;
  if (quality === "cinematic" || quality === "best" || quality === "ultra") return 96;
  if (effect.params.samples === "auto" || effect.params.samples === undefined || effect.params.samples === null) {
    return fallback;
  }
  return Math.max(2, Math.round(toFiniteNumber(effect.params.samples, fallback)));
};

const reconstructionWeight = (progress: number, curve: string) => {
  if (curve === "uniform" || curve === "linear") return 1;
  if (curve === "centerWeighted") {
    const distance = Math.abs(progress - 0.5) * 2;
    return Math.max(0.01, 1 - distance * 0.55);
  }
  const a0 = 0.35875;
  const a1 = 0.48829;
  const a2 = 0.14128;
  const a3 = 0.01168;
  const x = 2 * Math.PI * progress;
  return Math.max(0.001, a0 - a1 * Math.cos(x) + a2 * Math.cos(2 * x) - a3 * Math.cos(3 * x));
};

export const createMotionBlurSamplePlan = ({
  effect,
  layer,
  frameTime,
  fps,
  policy,
}: {
  effect: HyperFrameEffectInstance;
  layer: HyperFrameFrameDescriptorLayer;
  frameTime: number;
  fps: number;
  policy: FrameQualityPolicy;
}): MotionBlurSamplePlan => {
  const safeFps = Math.max(1, fps);
  const frameDuration = 1 / safeFps;
  const shutterAngle = clamp(toFiniteNumber(effect.params.shutterAngle, toFiniteNumber(effect.params.shutter, 1) * 180), 0, 1440);
  const shutterPhase = clamp(toFiniteNumber(effect.params.shutterPhase, -shutterAngle / 2), -720, 720);
  const amount = clamp(toFiniteNumber(effect.params.amount ?? effect.params.strength, 1), 0, 10);
  const shutterDuration = frameDuration * shutterAngle / 360 * amount;
  const centerTime = frameTime + frameDuration * shutterPhase / 360;
  const shutterStartTime = Math.max(0, centerTime - shutterDuration / 2);
  const shutterEndTime = Math.max(shutterStartTime, centerTime + shutterDuration / 2);
  const maxDimension = Math.max(layer.transform.width, layer.transform.height, 1);
  const angularSweep = Math.abs(layer.motion.angularVelocityDegrees) * shutterDuration;
  const translationSweep = layer.motion.speed * shutterDuration;
  const scaleSweep = Math.max(Math.abs(layer.motion.scaleVelocityX), Math.abs(layer.motion.scaleVelocityY)) * shutterDuration * maxDimension;
  const skewSweep = Math.max(Math.abs(layer.motion.skewVelocityX), Math.abs(layer.motion.skewVelocityY)) * shutterDuration * maxDimension * 0.35;
  const angularPixels = Math.abs(angularSweep) * Math.PI / 180 * maxDimension * 0.5;
  const estimatedMaxPixelDisplacement = translationSweep + angularPixels + scaleSweep + skewSweep;
  const estimatedEdgeVelocity = estimatedMaxPixelDisplacement / Math.max(0.0001, shutterDuration);
  const displacementSamples = Math.ceil(estimatedMaxPixelDisplacement / 1.35);
  const angularSamples = Math.ceil(Math.abs(angularSweep) / 18);
  const requiredSampleCount = Math.max(2, displacementSamples, angularSamples);
  const maxSamples = Math.max(2, Math.round(policy.maxSamples.motionBlur ?? 2));
  const fallbackSamples = Math.max(2, Math.round(policy.defaultSamples.motionBlur ?? maxSamples));
  const requestedSampleCount = requestedSamples(effect, fallbackSamples);
  const selectedSampleCount = Math.min(maxSamples, Math.max(requestedSampleCount, requiredSampleCount));
  const diagnostics: string[] = ["motion-blur-adaptive-samples-selected"];
  if (selectedSampleCount < requiredSampleCount) diagnostics.push("motion-blur-samples-clamped-by-backend");
  const curve = String(effect.params.sampleCurve || "filmic");
  const rawSamples = Array.from({ length: selectedSampleCount }, (_, index) => {
    const progress = selectedSampleCount === 1 ? 0.5 : index / (selectedSampleCount - 1);
    return {
      time: shutterStartTime + progress * (shutterEndTime - shutterStartTime),
      weight: reconstructionWeight(progress, curve),
    };
  });
  const totalWeight = Math.max(0.0001, rawSamples.reduce((total, sample) => total + sample.weight, 0));
  const qualityTier: MotionBlurQualityTier = selectedSampleCount < requiredSampleCount
    ? "clamped"
    : selectedSampleCount >= 96
      ? "cinematic"
      : selectedSampleCount >= 48
        ? "high"
        : selectedSampleCount >= 16
          ? "preview"
          : "draft";
  return {
    version: 2,
    shutterStartTime,
    shutterEndTime,
    sampleTimes: rawSamples.map((sample) => sample.time),
    weights: rawSamples.map((sample) => sample.weight / totalWeight),
    estimatedMaxPixelDisplacement,
    estimatedAngularSweep: angularSweep,
    estimatedEdgeVelocity,
    requiredSampleCount,
    selectedSampleCount,
    qualityTier,
    diagnostics,
  };
};
